beater/request/context.go

```diff
@@ -0,0 +1,165 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package request
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/apm-server/beater/headers"
+	logs "github.com/elastic/apm-server/log"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	mimeTypeAny             = "*/*"
+	mimeTypeApplicationJSON = "application/json"
+)
+
+var (
+	mimeTypesJSON = []string{mimeTypeAny, mimeTypeApplicationJSON}
+)
+
+// Context abstracts request and response information for http requests
+type Context struct {
+	Req *http.Request
+
+	statusCode int
+	err        interface{}
+	stacktrace string
+
+	w http.ResponseWriter
+}
+
+// Reset allows to reuse a context by removing all request specific information
+func (c *Context) Reset(w http.ResponseWriter, r *http.Request) {
+	c.Req = r
+
+	c.w = w
+	c.statusCode = http.StatusOK
+	c.err = ""
+	c.stacktrace = ""
+}
+
+// Header returns the http.Header of the context's writer
+func (c *Context) Header() http.Header {
+	return c.w.Header()
+}
+
+// StatusCode returns the context's status code
+func (c *Context) StatusCode() int {
```

Comment:
These getters feel non-idiomatic to me, are we worried about something specific here?

Reply:
You are right, it doesn't feel idiomatic. The intention was to avoid messing with the statusCode or Err outside the context. But I am also not satisfied with them, so changing for now and then we see how it comes together with the following changes.